lib/dynamoid/adapter_plugin/aws_sdk_v3/transact.rb

```diff
@@ -0,0 +1,158 @@
+          attributes = Dynamoid::Dumping.dump_attributes(attributes, model_class.attributes)
+          condition = "attribute_not_exists(#{model_class.hash_key})"
+          condition += " and attribute_not_exists(#{model_class.range_key})" if model_class.range_key? # needed?
+          {
+            put: {
+              item: sanitize_item(attributes),
+              table_name: model_class.table_name,
+              condition_expression: condition,
+            }
+          }
+        end
+
+        # Format the upsert action as a put. The previous document is completely overwritten if it already exists.
+        def transact_item_upsert(item)
+          model_class = item[:model_class]
+          raise "missing model_class" unless model_class
+          attributes = item[:attributes].clone
+          transact_add_timestamps(model_class, attributes) unless item[:skip_timestamps]
+          attributes[model_class.hash_key] ||= SecureRandom.uuid
+          attributes = Dynamoid::Dumping.dump_attributes(attributes, model_class.attributes)
+          raise "missing model_class" unless model_class
+          {
+            put: {
+              item: sanitize_item(attributes),
+              table_name: model_class.table_name,
+            }
+          }
+        end
+
+        # Format the delete action as a delete.
+        def transact_item_delete(item)
+          model_class = item[:model_class]
+          raise "missing model_class" unless model_class
+          attributes = Dynamoid::Dumping.dump_attributes(item[:attributes].clone, model_class.attributes)
+          hash_key = attributes.delete(model_class.hash_key)
+          key = {model_class.hash_key => hash_key}
+          key[model_class.range_key] = attributes.delete(model_class.range_key) if model_class.range_key?
+          # forces fast fail i.e. won't retry if record is missing
+          condition_expression = "attribute_exists(#{model_class.hash_key})"
+          condition_expression += " and attribute_exists(#{model_class.range_key})" if model_class.range_key? # needed?
+          {
+            delete: {
+              key: key,
+              table_name: model_class.table_name,
+              condition_expression: condition_expression
+            }
+          }
+        end
+
+        # Format the update action as an update.
+        # keys in item[:attributes] are NOT sanitized for injection attacks
+        def transact_item_update(item)
+          model_class = item[:model_class]
+          raise "missing model_class" unless model_class
+          attributes = item[:attributes].clone
+
+          # ensure types are correct for incoming attributes
+          all_types_string = attributes.keys.map{|k| [k,{type: :string}]}.to_h
+          transact_add_timestamps(model_class, attributes, skip_created_at: true) unless item[:skip_timestamps]
+          attributes.delete(model_class.inheritance_field) # do not update type field
```

Comment:
Are you wondering if ActiveRecord timestamp columns can be renamed? I don't believe they can be but that doesn't mean dynamoid couldn't :)
https://api.rubyonrails.org/classes/ActiveRecord/Timestamp.html

Reply:
I mean that in Rails it might be allowed to change values of "system" columns explicitly (timestamps, inheritance column, lock version column), e.g. with `user.update_attributes(type: 'Employee')`. But in the code above the `type` attribute is filtered out.

Reply:
Got it. I think I was seeing an error from dynamoid when trying to set the type but I also think we can remove that line so it shouldn't be an issue. Thanks.

Reply:
Removed this.